### PR TITLE
Bam tidy up

### LINKED ISSUE #269: BAM format sequence field parsing
Still combing through the BAM format specs. My lack of CS is going to show with this question, but for the `seq` field of the BAM format, as a 4-bit, high nibble first encoded read, does that mean that I have to make two passes through the array to construct the sequence string (e.g. high nibble transformation for each, then low nibble) or should I just do one pass, making serial transformation of each element (e.g. high nibble and low nibble for one, then continue)?

Or am I completely wrong. I am just puzzled is all...for example, I have an `l_seq` of 31, and therefore my `seq` array is `l_seq + 1// 2` long. That leaves my seq at a length of 16 if I do a high nibble only. Any guidance?

UPDATE: Nevermind. After messing around with samtools for comparison's sake, I found that to recreate the sequence it is high-nibble *then* low-nibble for each element of the array. Sorry for the interruption

### LINKED ISSUE #273: Clarification on BAM tag boundaries
The BAM spec has no description of how to determine the byte length of the H and Z tag values.
Is it the same as B where the first 4 bytes are the length?

This should also be clarified for the SAM spec.

Thanks

### PR DESCRIPTION
**Pull request by jkbonfield**
This incorporates the first commit from @daviesrob's BAMv2 PR to tidy up the BAM table, moving excessive footnotes into their own sections.  This coincidentally also fixes #273.

I then applied the same tidying logic to the new footnote on >65535 cigar operations and added clarification of the ordering in which sequence bases are encoded (fixes #269).

Finally, updated the version history as we forgot to do this when merging the long-cigar fix.

### CODE REVIEW

**Comment by innovate-invent** [comment]
[This line](https://github.com/jkbonfield/hts-specs/blob/feda1cc0a9889348d461d1445ae0e3d895cbac30/SAMv1.tex#L501) is the source of confusion for #273 
I am not sure how you would want to reword this to describe it as a ASCII representation of a HEX array.

**Comment by jkbonfield** [comment]
I've updated it, albeit minimally because I don't want the tables to be untidy unless we really need it.  Hopefully it's sufficient to imply that we really are talking about an ASCII string with 2 characters per byte.

File: SAMv1.tex [hunk 1]
```diff
@@ -903,19 +910,56 @@ \subsection{The BAM format}
...
-	The 4th byte, matching {\tt /\char94[cCsSiIf]\$/}, indicates the type of an element in the array.
-	Bytes from 5 to 8 encode a little-endian 32-bit integer which gives the number of elements in the array.
-	Bytes starting from the 9th store the array in the little-endian byte order; the number of these
-	bytes is determined by the type and the length of the array.}
+stored in the {\tt CG} tag and remove the now-redundant {\tt CG} tag.
+
+\subsubsection{SEQ encoding}\label{sec:seq}
+Sequence is encoded in 4-bit values, with adjacent bases packed into the same byte starting with the highest 4 bits first.
+For a sequence with an odd numbered length the bottom 4 bits of the last byte are undefined, but we recommend writing these as zero.
+The case-insensitive base codes `{\tt =ACMGRSVTWYHKDBN}' are mapped to $[0,15]$ respectively with all other characters mapping to `{\tt N}' (value 15).
```
[the remark this answers is not in the document]
Comment (jkbonfield) [reply SAMv1.tex]: Yes - I've made that mistake before infact thinking that 0 was N. :-)

`=` means this base matches the reference, and it's documented in both the SAM and BAM parts of the spec.

**Comment by jkbonfield** [comment]
@jmarshall:  This is 4 commits, but the first two are just formatting / text shifting.   If we agree on those but not the wording of the second two then I'm happy to squash and merge those two in and update this PR to be the second half only.

**Comment by jmarshall** [comment]
I very much agree with the direction of moving from footnotes to comprehensible text. But there's new text been written; it's not just been shifted.

However I don't find the table of allowed values and descriptions in the new §4.2.2 _Auxiliary data encoding_ as written very helpful: what this section needs is a discussion of the data's **representation** in BAM. And there is some important text in the footnotes that has not been carried over. I will write up what I think should be in this section over the weekend.

That's half of Rob's first commit; the other half is splitting up the conjoined fields in the table. Also :+1: from me, with suggestions: change **mq** in the first column to its proper **mapq** name now that it's not part of the shortened compound name; footnote 19 has fallen off the **n_cigar_op** entry (I realise this will be moot after the subsequent commit).

I haven't looked at the second two commits in detail yet, but my main initial concern with this when I first glanced at it was that IMHO adding 3 and 6 to footnote 9 about byte arrays makes it **less** clear. I see that's from one of the second two commits. To actually clarify it, we could rewrite it something like

> For example, the six-character Hex string '`1AE301`' represents the byte array [`0x1a`,`0xe3`,`0x1`].

**Comment by jmarshall** [comment]
@jkbonfield: I was going to suggest that you cherry-pick your _Add support for $>65535$ cigar operations. (Jul 2017)_ Version History addition and just apply it to **master** as that stuff has already landed. However such an entry highlights that at the time we deferred questions about bumping VN:1.5 to VN:1.6 for this cigar stuff, and it's now been quite a while and we still haven't come back to it.

So maybe you should create a PR just for that one line, and we can use it to argue about whether >64K cigars needs a version number bump and if so, what kind of version bump :smile:

**Comment by jkbonfield** [comment]
The "mq" name is there from before as it was "bin_mq_nl".  The new table clearly has MAPQ in the description. I assume Rob left the field as mq for continuity, but if we go down that road then should we also change l_read_name to l_qname too for consistency?  I'm OK with this either way really.

Regarding 4.2.2, I don't understand what you mean.  It does already have the BAM representation in there, albeit in C form which helps to clarify how it is derived.  I can't see information that was lost either.  It looks like a verbatim lift of the footnote to me.  Did you get around to rewriting this?

As for the second too commits, there were added specifically because someone found the initial text ambiguous, but you're correct that we should say 6-character HEX string instead of 6-byte to stress the difference.  I'll do some updates.

**Comment by yfarjoun** [comment]
👍

**Comment by jmarshall** [comment]
I've fixed a punctuation typo in the “Further clarification that hex encoding is ASCII text” commit and taken the liberty of adding my rewrite of the _Auxiliary data encoding_ section (which somewhere along the line became §4.2.4).

I trust that this makes it clear what I meant about spelling out the **representation** of the data!

IMHO what readers of the BAM-specific part of the spec want is **clarity about how the data is laid out in the file**, and to my mind this provides that significantly better than the existing footnotes or this PR's previous text. (Apologies for the delay.) The footnote information that was previously lost was the bit limiting the scope of “SAM always uses `i`” to “single (i.e., non-array) integer types”.

I've now fully reviewed the other new text in detail, and :+1: though I'd probably be more definitive in the SEQ encoding section than “we _recommend_ writing these as zero”.

So for me this is :+1: merge it :100:, and please do merge it if the other maintainers and @daviesrob agree that the last commit is an improvement.

**Comment by daviesrob** [comment]
Looks fine to me.

**Comment by jkbonfield** [comment]
Some great TeX vodoo there and I agree the rewrite is far more comprehensive and clearer.

Do we want to be explicit regarding the myriad choices of encoding SAM XX:i:10?  Eg something like this, after the 4, 5 or 7 bytes sentence: "The choice of sign and size to use for any integer value is implementation specific, although we encourage using the smallest size needed".  Specifically sign is something that differs between implementations, as XX:i:10 is equally justified as XXc<10> as it is XXC<10>.

File: SAMv1.tex [hunk 2]
```diff
@@ -921,48 +944,65 @@ \subsubsection{SEQ encoding}\label{sec:seq}
...
+than `{\tt H}' fields.}
+\begin{center}\small\byteboxsetup\begin{tabular}{l}
+\tagfield{Z}{\byteboxvector{1}{char}\bytebox{1}{\tt NUL}} \\
+\tagfield{H}{\byteboxvector{2}{hex hex}\bytebox{1}{\tt NUL}}
+\end{tabular}\end{center}
+
+\newcommand*{\arraytagfield}[3]{\tagfield{B}{\bytebox{1}{\tt #1}\bytebox{4}{\em count}\byteboxvector{#2}{#3}}}
+
+The representation of a `{\tt B}' array field starts with a sub-type character
+similar to the numeric field types above and an {\tt int32\_t} \emph{count}
```
[the remark this answers is not in the document]
Comment (jmarshall) [reply SAMv1.tex]: Read the commit message :smile:

(Also insert standard Java comments about unsigned here, which is presumably why _block_size_ is signed.)

**Comment by jmarshall** [comment]
Good idea about being explicit about the implementation-dependence of the cCsSiI choice. I've turned your suggested text into a footnote on the “chosen according to magnitude” bit.

File: SAMv1.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkbonfield) [reply SAMv1.tex]: Oops yes I didn't check the commit message.

I don't really think counts should ever be permitted to be signed.  Even if a language doesn't support unsigned types, it's just an implementation specific limit and not something that should affect a specification.

However I'm happy to ignore this problem for now as you're correct that it's part of a larger issue affecting more than this one field.

**Comment by jkbonfield** [comment]
@yfarjoun you approved this before and I'm assuming you're still OK, although @jmarshall's changes have arrived since then.  His aux tag text is a huge improvement of what we had though so plan on merging Monday unless I hear otherwise.

(I'll squash commits together by author, so 5 down to 3.)

**Comment by yfarjoun** [comment]
👍

**Comment by jkbonfield** [comment]
Squashed and merged as 9ffa426.

Anyone care greatly about updating the PDF, or wait until we merge a couple other issues which in theory are imminent.   SA tag clarification I think is ready, pending approval, and the CB/CR/CY one is minuted as to merge with @yfarjoun wording (assigned to Yossi, but I'll merge if you wish).

**Comment by jmarshall** [comment]
Yep, :+1: to batching up the PDF updates.
